### ISSUE TITLE
timer_remove does not stop timer

### ISSUE THREAD

**Issue by honzaskypala** [obsproject/obs-studio#8290]
### Operating System Info

macOS 13

### OBS Studio Version

29.0.1

### OBS Studio Log URL

https://obsproject.com/logs/BU2JOyggR8MBJ1rE

### Expected Behavior

With OBS scripting, `timer_remove` function should stop the timer, according to the documentation

### Current Behavior

Calling timer_remove does not stop the timer, the callback still continues running.

### Steps to Reproduce

Here is a simple Python script. If you load it to OBS and press the Start button, it will start calling the callback and it will print ticks into the log. Anyway, when you press Stop button, nothing changes, the ticks still continue to be written to the log.

```
import obspython as S

INTERVAL = 1

class RepeatAction:
    def __init__(self):
        self.timer_on = False
    
    def start_pressed(self, props, prop):
        if not self.timer_on:
            self.timer_on = True
            S.timer_add(self.ticker, INTERVAL * 1000)
    
    def stop_pressed(self, props, prop):
        if self.timer_on:
            self.timer_on = False
            S.timer_remove(self.ticker)

    def ticker(self):
        print("tick")

ra = RepeatAction()

def script_properties():  # ui
    props = S.obs_properties_create()
    S.obs_properties_add_button(props, "startButton", "Start", lambda x,y: RepeatAction.start_pressed(ra, x, y))
    S.obs_properties_add_button(props, "stopButton", "Stop", lambda x,y: RepeatAction.stop_pressed(ra, x, y))
    return props
```

**Comment by CodeYan01** [comment]
Please check your log, your script has an error.

**Comment by honzaskypala** [comment]
This error was from previous load of the script. You can notice in the log the script was loaded later second time, not producing the error anymore.

`10:58:58.175: [obs-scripting]: Loaded python script: timer-remove-bug.py`

**Comment by CodeYan01** [comment]
So, I tried adding print statements like this:
```py
def stop_pressed(self, props, prop):
        print('stop')
        if self.timer_on:
            print('stopped')
            self.timer_on = False
            S.timer_remove(self.ticker)```
and sure enough, both print statements worked, proving that it's not the if condition (which ofc looks fine), but the timer is indeed still running.

also changed the statements at the bottom to:
```py
    S.obs_properties_add_button(props, "startButton", "Start", lambda x,y: ra.start_pressed(x, y))
    S.obs_properties_add_button(props, "stopButton", "Stop", lambda x,y: ra.stop_pressed(x, y))
``` instead of using RepeatAction..., but same results.

Finally changed the ticker function to be a module-level function, not a class method,
```py
def ticker():
    print("tick")
```
and it worked, timer_remove did indeed stop it.

This shows that a possible problem is that timer_remove can't find the class function, as opposed to the global function. One difference with a class method and a module-level function is that the class method is passed the self parameter, whereas you usually can't timer_add a function with parameters. Perhaps this difference was not accounted for by timer_add and timer_remove code.

Tested on OBS 28.1.2, Python 3.9. I think it's safe to say this is something worth investigating.

**Comment by PatTheMav** [comment]
The root cause seems to be that when `timer_remove` is called, the referenced function object cannot be found - we iterate over all python callback functions registered with `libos`, but none match the reference given by `self.ticker`. The call becomes a no-op, and hence the timer is never cleared.

I'm not sure yet why that is, but that also explains why using a module function will work, as that function reference seems to be "stable".

**Comment by PatTheMav** [comment]
Alright, I've dug a bit deeper into this, and my conclusion (for now) is that this works as implemented (though not necessarily as intended):

References to instance functions are not "stable" as far as memory addresses or references are concerned - Python creates dynamic "descriptor" objects, which _can_ reuse the same memory address as before (Python "reuses" allocated memory as long as the last object has a reference count of `0`), but it's not guaranteed.

This means that the `PyObject` representing the bound method for `self.ticker` can change dynamically, but obs uses that reference to identify the timer - when you click "stop", Python gives us potentially an entirely different object reference.

TL;DR: Using instance methods for timers is not possible with how timers are implemented, always use module methods.
